From 821a85ba0d96ce5f53a15d8a538de268b891b3d0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Johannes=20Sch=C3=B6nberger?= <joschonb@microsoft.com>
Date: Wed, 1 Mar 2023 22:10:07 +0100
Subject: [PATCH] Remove unused SIFT GPU cuda texture reference

---
 lib/SiftGPU/ProgramCU.cu | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/lib/SiftGPU/ProgramCU.cu b/lib/SiftGPU/ProgramCU.cu
index 51c781341..9d842770d 100644
--- a/lib/SiftGPU/ProgramCU.cu
+++ b/lib/SiftGPU/ProgramCU.cu
@@ -1683,9 +1683,6 @@ void ProgramCU::MultiplyDescriptorG(CuTexImage* des1, CuTexImage* des2,
 												MatH, hdistmax, MatF, fdistmax);
 }
 
-
-texture<int,  1, cudaReadModeElementType> texDOT;
-
 #define ROWMATCH_BLOCK_WIDTH 32
 #define ROWMATCH_BLOCK_HEIGHT 1
 
@@ -1755,15 +1752,12 @@ void ProgramCU::GetRowMatch(CuTexImage* texDot, CuTexImage* texMatch, float dist
 	int num2 = texDot->GetImgWidth();
 	dim3 grid(1, num1/ROWMATCH_BLOCK_HEIGHT);
 	dim3 block(ROWMATCH_BLOCK_WIDTH, ROWMATCH_BLOCK_HEIGHT);
-	// texDot->BindTexture(texDOT);
 	RowMatch_Kernel<<<grid, block>>>((int*)texDot->_cuData,
 		(int*)texMatch->_cuData, num2, distmax, ratiomax);
 }
 
 #define COLMATCH_BLOCK_WIDTH 32
 
-//texture<int3,  1, cudaReadModeElementType> texCT;
-
 void __global__  ColMatch_Kernel(int3*d_crt, int* d_result, int height, int num2, float distmax, float ratiomax)
 {
 	int col = COLMATCH_BLOCK_WIDTH * blockIdx.x + threadIdx.x;
