diff --git a/cmake/FindLZ4.cmake b/cmake/FindLZ4.cmake
index 9f9d5232c..d3abd598b 100644
--- a/cmake/FindLZ4.cmake
+++ b/cmake/FindLZ4.cmake
@@ -42,47 +42,22 @@
 #                              search for LZ4 includes.
 # LZ4_LIBRARY_DIR_HINTS: List of additional directories in which to
 #                              search for LZ4 libraries.
+find_path(LZ4_INCLUDE_DIR
+  NAMES lz4.h
+  DOC "lz4 include directory")
+mark_as_advanced(LZ4_INCLUDE_DIR)
+find_library(LZ4_LIBRARY
+  NAMES lz4 liblz4
+  DOC "lz4 library")
+mark_as_advanced(LZ4_LIBRARY)
 
-set(LZ4_INCLUDE_DIR_HINTS "" CACHE PATH "LZ4 include directory")
-set(LZ4_LIBRARY_DIR_HINTS "" CACHE PATH "LZ4 library directory")
-
-unset(LZ4_FOUND)
-unset(LZ4_INCLUDE_DIRS)
-unset(LZ4_LIBRARIES)
-
-list(APPEND LZ4_CHECK_INCLUDE_DIRS
-    ${LZ4_INCLUDE_DIR_HINTS}
-    /usr/include
-    /usr/local/include
-    /opt/include
-    /opt/local/include
-)
-
-list(APPEND LZ4_CHECK_LIBRARY_DIRS
-    ${LZ4_LIBRARY_DIR_HINTS}
-    /usr/lib
-    /usr/local/lib
-    /usr/lib/x86_64-linux-gnu
-    /opt/lib
-    /opt/local/lib
-)
-
-find_path(LZ4_INCLUDE_DIRS
-    NAMES
-    lz4.h
-    PATHS
-    ${LZ4_CHECK_INCLUDE_DIRS})
-find_library(LZ4_LIBRARIES
-    NAMES
-    lz4
-    PATHS
-    ${LZ4_CHECK_LIBRARY_DIRS})
-
-if(LZ4_INCLUDE_DIRS AND LZ4_LIBRARIES)
-    set(LZ4_FOUND TRUE)
-endif()
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(LZ4
+  REQUIRED_VARS LZ4_LIBRARY LZ4_INCLUDE_DIR)
 
 if(LZ4_FOUND)
+    set(LZ4_INCLUDE_DIRS "${LZ4_INCLUDE_DIR}")
+    set(LZ4_LIBRARIES "${LZ4_LIBRARY}")
     message(STATUS "Found LZ4")
     message(STATUS "  Includes : ${LZ4_INCLUDE_DIRS}")
     message(STATUS "  Libraries : ${LZ4_LIBRARIES}")
