diff --git a/lib/PoissonRecon/CMakeLists.txt b/lib/PoissonRecon/CMakeLists.txt
index cedb75151..491baa77a 100644
--- a/lib/PoissonRecon/CMakeLists.txt
+++ b/lib/PoissonRecon/CMakeLists.txt
@@ -1,5 +1,5 @@
 if(NOT IS_MSVC)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops -ffast-math -std=c++11")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops -ffast-math")
 endif()
 
 add_definitions("-DRELEASE")
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 7d5049253..649ce594b 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -29,24 +29,6 @@
 #
 # Author: Johannes L. Schoenberger (jsch-at-demuc-dot-de)
 
-if(IS_MSVC)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W3")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W3")
-elseif(IS_GNU OR IS_CLANG)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
-    # We want c++14 mainly for supporting latest dependencies, but older GCC
-    # versions in combination with CUDA do not yet support it. In the worst
-    # case, this will lead to compile errors and there is not much we can
-    # do on the COLMAP side.
-    if(CUDA_ENABLED AND
-       IS_GNU AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 6)
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
-    else()
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
-    endif()
-endif()
-
 add_subdirectory(base)
 add_subdirectory(controllers)
 add_subdirectory(estimators)
