diff --git a/CMakeLists.txt b/CMakeLists.txt
old mode 100755
new mode 100644
index 7886d36..3f57631
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -48,7 +48,7 @@ include(CheckCXXCompilerFlag)
 # Include helper macros and commands, and allow the included file to override
 # the CMake policies in this file
 include(${CMAKE_CURRENT_SOURCE_DIR}/cmake/CMakeHelper.cmake NO_POLICY_SCOPE)
-
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
 
 ################################################################################
 # Options
diff --git a/lib/PoissonRecon/CMakeLists.txt b/lib/PoissonRecon/CMakeLists.txt
index 7841eb6..9f27293 100644
--- a/lib/PoissonRecon/CMakeLists.txt
+++ b/lib/PoissonRecon/CMakeLists.txt
@@ -1,5 +1,5 @@
 if(NOT IS_MSVC)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops -ffast-math -std=c++11")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -funroll-loops -ffast-math")
 endif()
 
 add_definitions("-DRELEASE")
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index e1821cd..5d8ba0b 100755
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -29,24 +29,6 @@
 #
 # Author: Johannes L. Schoenberger (jsch-at-demuc-dot-de)
 
-if(IS_MSVC)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W3")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W3")
-elseif(IS_GNU OR IS_CLANG)
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall")
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall")
-    # We want c++14 mainly for supporting latest dependencies, but older GCC
-    # versions in combination with CUDA do not yet support it. In the worst
-    # case, this will lead to compile errors and there is not much we can
-    # do on the COLMAP side.
-    if(CUDA_ENABLED AND
-       IS_GNU AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 6)
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
-    else()
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
-    endif()
-endif()
-
 add_subdirectory(base)
 add_subdirectory(controllers)
 add_subdirectory(estimators)
