diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -93,7 +93,6 @@ find_package(Boost REQUIRED COMPONENTS
              program_options
              filesystem
              graph
-             system
              unit_test_framework)
 
 find_package(Eigen3 CONFIG REQUIRED)
@@ -369,7 +368,6 @@ set(COLMAP_EXTERNAL_LIBRARIES
     ${CMAKE_DL_LIBS}
     ${Boost_FILESYSTEM_LIBRARY}
     ${Boost_PROGRAM_OPTIONS_LIBRARY}
-    ${Boost_SYSTEM_LIBRARY}
     Eigen3::Eigen
     ${GLOG_LIBRARIES}
     ${FLANN_LIBRARIES}
diff --git a/cmake/CMakeConfig.cmake.in b/cmake/CMakeConfig.cmake.in
--- a/cmake/CMakeConfig.cmake.in
+++ b/cmake/CMakeConfig.cmake.in
@@ -81,7 +81,6 @@ if(COLMAP_FIND_QUIETLY)
     find_package(Boost COMPONENTS
                 program_options
                 filesystem
-                system
                 unit_test_framework
                 QUIET)
 
@@ -105,7 +104,6 @@ else()
     find_package(Boost COMPONENTS
                 program_options
                 filesystem
-                system
                 unit_test_framework
                 REQUIRED)
